File: DS/BFS.cpp
#include<bits/stdc++.h>

using namespace std;

vector<bool> visited;

void add_edge(vector<int> g[], int v1, int v2){
    g[v1].push_back(v2);
    g[v2].push_back(v1);
}

void bfs(vector<int> g[], int n){
    queue<int> q;
    q.push(n);
    visited[n] = true;

    while(!q.empty()){
        int f = q.front();
        q.pop();

        cout << f << " ";
        
        for(auto i = g[f].begin(); i != g[f].end(); i++){
            if(!visited[*i]){
                q.push(*i);
                visited[*i] = true;
            }
                
        }
    }

}


int main(){
    int n, m;
    cout << "Enter the number of vertices and Edges :- " << " ";
    cin >> n >> m;
    cout << "\n";
    vector<int> g[n];
    cout << "Enter the pair of vertices :- " << endl;
    int a, b;
    for(int i=0; i<m; i++){
        cin >> a >> b;
        add_edge(g, a, b);
    }

    visited.assign(n, false);

    for (int i=0; i<n; i++){
        if(!visited[i])
            bfs(g, i);
    }
    return 0;
}

File: DS/BFS.py
from collections import defaultdict

class Graph:

    def __init__(self, size):
        self.size = size
        self.graph = defaultdict(list)
        self.visited = [False] * size

    
    def add_edge(self, src, dest):
        self.graph[src] = dest
        self.graph[dest] = src

    
    def bfs(self, start):

        queue = list()
        queue.append(start)
        
        self.visited[start] = True

        while queue:
            
            temp = queue.pop(0)
            print(temp, end=' ')

            for i in self.graph[temp]:
                if not self.visited[i]:
                    queue.append(i)
                    self.visited[i] = True


    def bft(self):
        for i in range(self.size):
            if not self.visited[i]:
                self.bfs(i)

    def print_graph(self):
        for i in range(self.size):
            temp = self.graph[i]
            print("Adjacency list of vertex {}\n head ".format(i), end=' ')
            for j in temp:
                print(j, end=' ')
            print()


n, m = map(int, input('Enter the number of vertices and Edges :- ').split())
graph = Graph(n)
print("Enter vertices pairs :- ")
for _ in range(m):
    a, b = map(int, input().split())
    graph.add_edge(a, b)

graph.print_graph()
graph.bfs(0)
            
